website_canonical_url/models/website.py

```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 Simone Orsi <simone.orsi@camptocamp.com>
+# Copyright initOS GmbH 2016
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, api, fields, exceptions, _
+from odoo.http import request
+from urlparse import urlparse, urljoin
+
+
+class Website(models.Model):
+    _inherit = 'website'
+
+    canonical_domain = fields.Char(
+        help='Canonical domain is used to build unique canonical URLs '
+             'to make SEO happy.'
+    )
+
+    @api.constrains('canonical_domain')
+    def _check_canonical_domain(self):
+        domain = self.canonical_domain
+        if domain and not urlparse(domain).scheme:
+            raise exceptions.ValidationError(_(
+                'Canonical domain must contain protocol `http(s)://`'
+            ))
+
+    @api.multi
+    def get_canonical_url(self, req=None):
+        return urljoin(
+            self._get_canonical_domain(),
+            self._get_canonical_relative_url(req=req)
+        )
+
+    @api.multi
+    def _get_canonical_domain(self):
+        self.ensure_one()
+        if self.canonical_domain:
+            return self.canonical_domain
+        params = self.env['ir.config_parameter'].sudo()
+        return params.get_param('web.base.url')
+
+    @api.multi
+    def _get_canonical_relative_url(self, req=None):
+        req = req or request
+        parsed = urlparse(req.httprequest.path)
+        # here we get the full path and qstring args stripped out already
+        canonical_url = parsed.path
+        # build lang path if request lang does not match default lang
+        lang_path = ''
+        if req.lang != req.website.default_lang_code:
+            # eg: language is `it_IT` but main lang is `en_US`
+            lang_path = '/%s/' % req.lang
+        # handle special case for rerouted requests to root path
+        if self._is_root_page(canonical_url, lang_path=lang_path):
```

Comment:
So, after all, leave this implementation as it was, which was already working fine (except for the missing hostname) and taking into account redirections and reroutings:

https://github.com/OCA/website/blob/6ec9b1598a9ad6f92186f338562fbdbd336d3439/website_canonical_url/models/website.py#L14-L35

If you really want to improve it, I'd better concentrate in the root page detection algorithm, which should be exactly the same as upstream if we don't want to hit some corner cases, but nothing else.